### PR TITLE
Touch wrapper to allow clicking by image

### PR DESCRIPTION
**Pull request by jsa34**
Attempting to create command wrapper, but not seeming to work - very novice in JS, so pointers welcome!

### CODE REVIEW

**Comment by jsa34** [comment]
I am trying to debug this, as it seems to accept the request and thinks it has clicked, but I am trying to ascertain if the wrong coordinates have been found or if the click is failing to happen.... I return the coordinates to debug but not sure where the coordinates reference - using the Java Point coordinates of the screen, the location appears to be wrong for what is returned... it is locating a random space that I wouldn't imagine the image recognition would locate as it's empty white space.

File: lib/commands/touch.js [hunk 1]
```diff
@@ -0,0 +1,33 @@
+import log from '../logger';
+
+let commands = {}, extensions = {};
+
+commands.performActions = async function (actions) {
```
Comment (mykola-mokhnach) [inline lib/commands/touch.js]: there is no need for preprocessing. `return await this.winAppDriver.sendCommand('/actions', 'POST', {actions});` would be enough

File: lib/commands/touch.js [hunk 2]
```diff
@@ -0,0 +1,33 @@
+import log from '../logger';
+
+let commands = {}, extensions = {};
```
Comment (mykola-mokhnach) [inline lib/commands/touch.js]: extensions are not needed. keep only commands

**Comment by jsa34** [comment]
Updated as suggested, but the reason I was adding the preprocessing was because I get an error saying:
[debug] [WD Proxy] Matched '/actions' to command name 'performActions'
[debug] [WD Proxy] Proxying [POST /actions] to [POST http://127.0.0.1:4724/wd/hub/session/493D6DFB-AF86-4FD9-9B39-6D62119FCE2A/actions] with body: {"actions":[{"type":"pointer","id":"mouse","parameters":{"pointerType":"touch"},"actions":[{"type":"pointerMove","x":105.5,"y":334,"duration":0},{"type":"pointerDown"},{"type":"pause","duration":125},{"type":"pointerUp"}]}]}
[WinAppDriver] [STDOUT]
[WinAppDriver] [STDOUT]
[WinAppDriver] [STDOUT] ==========================================
[WinAppDriver] [STDOUT] POST /wd/hub/session/493D6DFB-AF86-4FD9-9B39-6D62119FCE2A/actions HTTP/1.1
[WinAppDriver] [STDOUT] Accept: application/json, */*
[WinAppDriver] [STDOUT] Connection: close
[WinAppDriver] [STDOUT] Content-Length: 224
[WinAppDriver] [STDOUT] Content-Type: application/json; charset=utf-8
[WinAppDriver] [STDOUT] Host: 127.0.0.1:4724
[WinAppDriver] [STDOUT] User-Agent: appium
[WinAppDriver] [STDOUT]
[WinAppDriver] [STDOUT] {"actions":[{"type":"pointer","id":"mouse","parameters":{"pointerType":"touch"},"actions":[{"type":"pointerMove","x":105.5,"y":334,"duration":0},{"type":"pointerDown"},{"type":"pause","duration":125},{"type":"pointerUp"}]}]}
[WinAppDriver] [STDOUT] HTTP/1.1 400 Bad Request
[WinAppDriver] [STDOUT] Content-Length: 168
[WinAppDriver] [STDOUT] Content-Type: application/json
[WinAppDriver] [STDOUT]
[WinAppDriver] [STDOUT] {"status":100,"value":{"error":"invalid argument","message":"\"button\" in a pointer action JSON payload is undefined or is not an Integer greater than or equal to 0"}}
[WD Proxy] Got response with status 400: {"status":100,"value":{"error":"invalid argument","message":"\"button\" in a pointer action JSON payload is undefined or is not an Integer greater than or equal to 0"}}
[debug] [W3C] Matched W3C error code 'invalid argument' to InvalidArgumentError
[debug] [MJSONWP (db69288b)] Encountered internal error running command: InvalidArgumentError: "button" in a pointer action JSON payload is undefined or is not an Integer greater than or equal to 0
[debug] [MJSONWP (db69288b)]     at errorFromW3CJsonCode (C:\Users\automation\Documents\local_modules\appium\node_modules\appium-base-driver\lib\protocol\errors.js:780:25)
[debug] [MJSONWP (db69288b)]     at ProxyRequestError.getActualError (C:\Users\automation\Documents\local_modules\appium\node_modules\appium-base-driver\lib\protocol\errors.js:663:14)
[debug] [MJSONWP (db69288b)]     at JWProxy.command (C:\Users\automation\Documents\local_modules\appium\node_modules\appium-base-driver\lib\jsonwp-proxy\proxy.js:264:19)
[HTTP] <-- POST /wd/hub/session/db69288b-66d5-4b90-806d-5367e6f25a30/element/appium-image-element-1e0573c3-2eb6-4c63-9922-84dc46187293/click 500 1199 ms - 191

**Comment by mykola-mokhnach** [comment]
> Updated as suggested, but the reason I was adding the preprocessing was because I get an error suggesting button is missing in the payload saying

Perhaps this only means it is necessary to adjust the touch action inside the base driver and add a button to it

**Comment by mykola-mokhnach** [comment]
Could you please send the following chain there and check if windows server handles it:

[WinAppDriver] [STDOUT] {"actions":[{"type":"pointer","id":"mouse","parameters":{"pointerType":"touch"},"actions":[{"type":"pointerMove","x":105.5,"y":334,"duration":0},{"type":"pointerDown", "button": 0},{"type":"pause","duration":125},{"type":"pointerUp", "button": 0}]}]}

**Comment by mykola-mokhnach** [comment]
https://github.com/appium/appium-base-driver/blob/d0ac39d8fc8e09495df128e98b002c7f9c9d7899/lib/basedriver/image-element.js#L177

**Comment by jsa34** [comment]
Added back in the function for adding the button value to payload and it does indeed appear to click and work now, albeit clicking on the wrong element... Not sure if this is because the image matching has picked up the wrong element, need to see how to customise the image location settings ...

**Comment by jsa34** [comment]
The following chain was successful, so I suggest the modification to base driver would solve this issue - not sure about how to customise the image locating though, as the screenshot I took was at native resolution (no scaling) but still seemed to find the wrong location match...

{"actions":[{"type":"pointer","id":"mouse","parameters":{"pointerType":"touch"},"actions":[{"type":"pointerMove","x":62.5,"y":494.5,"duration":0},{"type":"pointerDown","button":0},{"type":"pause","duration":125},{"type":"pointerUp","button":0}]}]}

File: lib/commands/touch.js [hunk 3]
```diff
@@ -1,4 +1,4 @@
-Let commands = {};
+let commands = {};
```
Comment (mykola-mokhnach) [inline lib/commands/touch.js]: const

**Comment by mykola-mokhnach** [comment]
Please bump the minimum base-driver version in package.json to 5.4.3

**Comment by mykola-mokhnach** [comment]
It looks like the linter is not very happy: https://travis-ci.org/github/appium/appium-windows-driver/jobs/664091420#L329

**Comment by mykola-mokhnach** [comment]
`npm run lint` could be executed locally on the package to check linting before committing it

**Comment by jsa34** [comment]
Apologies - was having to use the online text editor for the moment and thought it would make it easier to get the quick changes committed, but apparently not!  Will wait until I can get to PC

**Comment by jsa34** [comment]
Thank you for your patience, @mykola-mokhnach !

File: lib/commands/touch.js [hunk 4]
```diff
@@ -1,7 +1,8 @@
 const commands = {};
 
-commands.performActions = async function (actions) {  
-  return await this.winAppDriver.sendCommand('/actions', 'POST', {actions});};
```
Comment (mykola-mokhnach) [inline lib/commands/touch.js]: it makes sense to add a comment here that this method is only needed to make clicks on `-image` elements working properly

Comment (jsa34) [reply lib/commands/touch.js]: Added

**Pull request merged**